### PR TITLE
arch: x86_64: Copy cache related information into guest CPUID

### PR DESCRIPTION
**Pull request by likebreath**
It includes the TLB/Cache/Prefetch information from CPUID leaf 0x2,
Deterministic Cache Parameter information from CPUID leaf 0x4 (with
all its valid sub-leaves), and additional cache-related information from
CPUID leaf 0x8000_0006.

Fixes: #1769

Signed-off-by: Bo Chen <chen.bo@intel.com>

### CODE REVIEW

File: arch/src/x86_64/mod.rs [hunk 1]
```diff
@@ -394,6 +394,65 @@ pub fn configure_vcpu(
...
+
+            break;
+        }
+
+        cpuid
+            .push(CpuIdEntry {
+                function: 0x4,
+                index: subleaf_idx,
+                // Todo: ignoring the informaiton related to addressable processor IDs
+                eax: leaf.eax & 0xff,
```
Comment (likebreath) [inline arch/src/x86_64/mod.rs:442]: @rbradford This PR is a draft, as the guest kernel doesn't boot when the CPUID 0x4 is copied to the guest in this PR. I will look into it further and will ask your inputs tomorrow if I can't resolve it.

The guest kernel simply hangs after the kernel prints `[    0.004000] Mountpoint-cache hash table entries: 2048 (order: 2, 16384 bytes)`. The complete log is [here](https://gist.github.com/likebreath/a68b1be7df6968e7e7d4d9f5bd6eabc3). My guess was the bug might be related to the sub-leaf value on `EAX` for bits 14-31 which are related to the addressable processor IDs. That information depend on available vcpus and NUMA nodes for the guest and should not be copied directly from the host. I tried to ignore the values from host and simply set them as 0s, but it does not work either. 

Quotes from the [SDM](https://www.intel.com/content/www/us/en/architecture-and-technology/64-ia-32-architectures-software-developer-vol-2a-manual.html) (page 293-294):

```
Bits 25 - 14: Maximum number of addressable IDs for logical processors sharing this cache**, ***.
Bits 31 - 26: Maximum number of addressable IDs for processor cores in the physical 
```

Comment (rbradford) [reply arch/src/x86_64/mod.rs:442]: You need to populate it with the details from the `CpuTopology` struct.

**Comment by likebreath** [comment]
I figured the reason why my original patch was causing guest OS hanging with booting: the CPUID entries need have the right `flags` ( set to `CPUID_FLAG_VALID_INDEX`). It fixed the hanging on my dev machine, and let's see how the CI works.

**Comment by likebreath** [comment]
@michael2012z @MrXinWang The AArch64 CI machine is experiencing the `ERROR: Failed to clean the workspace` issue. Would you please manually cleanup the CI machine? Thank you.

**Comment by MrXinWang** [comment]
@likebreath  Done

**Comment by rbradford** [comment]
@likebreath I worried we might not even have needed to do this. Did you check if the values were wrong inside the guest?. I had been using lscpu but I think the cache values it quotes are cumulative.

All the tests I run below are from master branch

e.g. `--cpus boot=4,topology=2:2:1:1`
```
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              2
Core(s) per socket:              2
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           78
Model name:                      Intel(R) Core(TM) i5-6260U CPU @ 1.80GHz
Stepping:                        3
CPU MHz:                         1799.990
BogoMIPS:                        3599.98
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       64 KiB
L1i cache:                       64 KiB
L2 cache:                        512 KiB
L3 cache:                        4 MiB
```

From host:
```
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              2
Core(s) per socket:              2
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           78
Model name:                      Intel(R) Core(TM) i5-6260U CPU @ 1.80GHz
Stepping:                        3
CPU MHz:                         700.043
CPU max MHz:                     2900.0000
CPU min MHz:                     400.0000
BogoMIPS:                        3600.00
Virtualization:                  VT-x
L1d cache:                       64 KiB
L1i cache:                       64 KiB
L2 cache:                        512 KiB
L3 cache:                        4 MiB
```

From a VM with a single CPU:
```
CPU(s):                          1
On-line CPU(s) list:             0
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           78
Model name:                      Intel(R) Core(TM) i5-6260U CPU @ 1.80GHz
Stepping:                        3
CPU MHz:                         1799.990
BogoMIPS:                        3599.98
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        256 KiB
L3 cache:                        4 MiB
```

**Comment by likebreath** [comment]
> I worried we might not even have needed to do this. Did you check if the values were wrong inside the guest?. I had been using lscpu but I think the cache values it quotes are cumulative.

@rbradford I had the same question and I observed the same on my side with my own machine. But I assumed the observation is platform dependent given  #1769 reported wrong L1/L2 cache sizes (based your setup). 

Looks like the issue has been fixed also on your side. So maybe this cache size issue is not a problem anymore?

**Pull request closed without merging**